### ISSUE TITLE
Fix S4457 - false positive with local functions

### ISSUE THREAD

**Issue by dnperfors** [SonarSource/sonar-csharp#1380]
### Description

The S4457 rule gives a false positive on an iterator method where the iterator itself is implemented in a local function.

### Repro steps

```
private static IEnumerable<SyntaxTree> CompileSources(IEnumerable<string> sources)
{
    if (sources == null) throw new ArgumentNullException(nameof(sources));
    IEnumerable<SyntaxTree> Iterator()
    {
        foreach (string source in sources)
        {
            yield return CSharpSyntaxTree.ParseText(source);
        }
    }
    return Iterator();
}
```

### Expected behavior

No error in this code

### Actual behavior

The rule s4457 saying that I should split this method

### Known workarounds

Move the local function to a private method

### Related information

* SonarC# Version: ?
* Visual Studio Version: VS15.6.7
* SonarLint Version: 4.0.0.3479 (
* If running through the Scanner for MSBuild, its version: using the Visual Studio plugin

**Comment by MaxTranced** [comment]
I also get a false positive for this code:

```
        internal static async Task<OwinHttpContext> ToHttpInterfacesAsync(
            IOwinContext context,
            AuthenticationProperties properties,
            ConnectisAuthenticationOptions options)
        {
            if (context == null)
            {
                throw new ArgumentNullException("context");
            }

            IFormCollection formData = null;
            if (context.Request.Body != null)
            {
                formData = await context.Request.ReadFormAsync().ConfigureAwait(false);
            }

            return new OwinHttpContext(context, formData, properties, options);
        }
```

**Comment by valhristov** [comment]
@dnperfors our C# analyzer does not correctly recognize C# 7+ syntax yet and this looks like the reason for the FP.

**Comment by valhristov** [comment]
@MaxTranced your code does not look like a FP to me. As far as I understand, the argument check will happen only if you await the `ToHttpInterfacesAsync` method. The idea of the rule is to enforce the argument exceptions to be thrown without the need to `await`.

**Comment by MaxTranced** [comment]
> @MaxTranced your code does not look like a FP to me. As far as I understand, the argument check will happen only if you await the ToHttpInterfacesAsync method. The idea of the rule is to enforce the argument exceptions to be thrown without the need to await:

OK then, that makes sense... Maybe then the text of the rule could be updated to no longer mention an iterator and to properly explain the reasoning. Basically very similar to your explanation, which makes much more sense.

**Comment by dnperfors** [comment]
> @dnperfors our C# analyzer does not correctly recognize C# 7+ syntax yet and this looks like the reason for the FP.

That sounds logical, are there any plans to support C# 7+?

**Comment by valhristov** [comment]
@dnperfors yes, we do want to support C# 7, but it is a difficult task. Hopefully we will be able to deliver support this year, but I cannot commit to any date at this moment.

@MaxTranced I created a ticket #1387, we will update the rule with the next release.

**Comment by Evangelink** [comment]
@valhristov @dnperfors We actually can workaround having the support for C# 7 in this case. PR to come.
